### PR TITLE
support canonical version only metadata validation

### PR DESCRIPTION
**Pull request by OsmondJiang**
[AB#230344](https://dev.azure.com/ceapex/d3d54af3-265a-4f18-95f6-9a46397ca583/_workitems/edit/230344)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/docfx/pull/6035)

### CODE REVIEW

File: src/docfx/lib/log/Error.cs [hunk 1]
```diff
@@ -47,12 +47,16 @@ public Error(ErrorLevel level, string code, string message, FilePath? file = nul
             Name = name;
         }
 
-        public Error WithCustomError(CustomError customError)
+        public Error WithCustomRule(CustomRule customRule, bool? isCanonicalVersion = null)
         {
+            var level = isCanonicalVersion != null && customRule.CanonicalVersionOnly
```
Comment (yufeih) [inline src/docfx/lib/log/Error.cs:52]: nit: consider split it into 2 lines, one is to apply `customRule.Severity` override, the other to apply `isCanonical` override.

File: src/docfx/lib/schema/JsonSchemaValidator.cs [hunk 2]
```diff
@@ -17,34 +18,43 @@ internal class JsonSchemaValidator
         private readonly JsonSchema _schema;
         private readonly JsonSchemaDefinition _definitions;
         private readonly MicrosoftGraphAccessor? _microsoftGraphAccessor;
-        private readonly ListBuilder<(JsonSchema schema, string key, JToken value, SourceInfo? source)> _metadataBuilder;
+        private readonly ListBuilder<(JsonSchema schema, string key, JToken value, SourceInfo? source, bool? isCanonicalVersion)> _metadataBuilder;
+        private static readonly ThreadLocal<bool?> _isCanonicalVersion = new ThreadLocal<bool?>();
```
Comment (yufeih) [inline src/docfx/lib/schema/JsonSchemaValidator.cs:22]: nit: use `t_` for thread local variables to tell the different between member variables and thread static variables.
```suggestion
        private static readonly ThreadLocal<bool?> t_isCanonicalVersion = new ThreadLocal<bool?>();
```

File: src/docfx/lib/log/Error.cs [hunk 3]
```diff
@@ -47,12 +47,16 @@ public Error(ErrorLevel level, string code, string message, FilePath? file = nul
             Name = name;
         }
 
-        public Error WithCustomError(CustomError customError)
+        public Error WithCustomRule(CustomRule customRule, bool? isCanonicalVersion = null)
```
Comment (yufeih) [inline src/docfx/lib/log/Error.cs:50]: consider promote `IsCanonicalVersion` into a property (similar to `Name`) so you don't need to use a tuple [here](https://github.com/dotnet/docfx/pull/6035/files#diff-eb444f860ada42127995887c1b00a00aR519)

File: src/docfx/lib/schema/JsonSchemaValidator.cs [hunk 4]
```diff
@@ -517,13 +527,19 @@ private void PostValidateDocsetUnique(List<Error> errors)
 
             foreach (var group in validatedMetadataGroups)
             {
-                if (group.Count() > 1)
+                IEnumerable<(JsonSchema schema, string key, JToken value, SourceInfo? source, bool? isCanonicalVersion)> items = group;
+                var (metadataValue, (metadataKey, _)) = group.Key;
+                if (CanonicalVersionOnly(metadataKey, Errors.JsonSchema.DuplicateAttributeCode))
```
Comment (yufeih) [inline src/docfx/lib/schema/JsonSchemaValidator.cs:532]: how about doing the filtering be done in line 523?

File: src/docfx/lib/log/Error.cs (hunk 3 above)
Comment (OsmondJiang) [reply src/docfx/lib/log/Error.cs:50]: 👍

**Pull request merged**